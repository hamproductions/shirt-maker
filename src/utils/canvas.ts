export const getTextDimensions = (ctx: CanvasRenderingContext2D, text: string) => {
  const size = ctx.measureText(text);
  return {
    width: Math.abs(size.actualBoundingBoxLeft) + Math.abs(size.actualBoundingBoxRight),
    height: Math.abs(size.actualBoundingBoxAscent) + Math.abs(size.actualBoundingBoxDescent)
  };
};
