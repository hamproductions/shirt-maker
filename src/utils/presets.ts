export type PresetData = {
  text?: string;
  size: number;
  placeholder?: { key: string; default: string };
}[][];
export type Preset = {
  text: string;
  data: PresetData;
};
export const PRESETS: Preset[] = [
  {
    text: '日本語は全て<oshi>から学んだ',
    data: [
      [{ text: '日本語は', size: 1070 }],
      [
        { text: '全て', size: 680 },
        { placeholder: { key: 'oshi', default: '女性声優' }, size: 680 }
      ],
      [{ text: 'から学んだ', size: 980 }]
    ]
  },
  {
    text: '日本語は全て<oshi>のため学んだ',
    data: [
      [{ text: '日本語は', size: 1070 }],
      [
        { text: '全て', size: 680 },
        { placeholder: { key: 'oshi', default: '女性声優' }, size: 680 }
      ],
      [{ text: 'のため学んだ', size: 800 }]
    ]
  },
  {
    text: '私が日本に来た理由は<oshi>を会うために',
    data: [
      [{ text: '私が日本に', size: 650 }],
      [{ text: '来た理由は', size: 800 }],
      [{ placeholder: { key: 'oshi', default: '女性声優' }, size: 700 }],
      [{ text: 'を会うために', size: 700 }]
    ]
  }
];
