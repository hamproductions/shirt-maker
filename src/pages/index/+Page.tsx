import { useState } from 'react';

import { useTranslation } from 'react-i18next';
import { FaCopy, FaDownload } from 'react-icons/fa6';
import { useToaster } from '../../context/ToasterContext';

import { Box, HStack, Stack } from 'styled-system/jsx';
import { Metadata } from '~/components/layout/Metadata';
import * as RadioGroup from '~/components/ui/styled/radio-group';
import { PRESETS } from '~/utils/presets';
import { Template } from '~/components/template/Template';
import { Input } from '~/components/ui/styled/input';
import { Button } from '~/components/ui/styled/button';

export function Page() {
  const { toast } = useToaster();
  const { t, i18n } = useTranslation();
  const [presetIndex, setPresetIndex] = useState(0);
  const [placeholderData, setPlaceholderData] = useState<Record<string, string>>({});

  const preset = PRESETS[presetIndex];
  const placeholders = PRESETS[presetIndex].data.flatMap((d) => d.filter((e) => !!e.placeholder));

  const title = 'Dumb Shirt Maker';

  const replacePlaceholder = (templateString: string, placeholderData: Record<string, string>) => {
    return templateString.replaceAll(/<(.*?)>/g, (_, key) => {
      return (
        placeholderData[key] ||
        placeholders.find((p) => p.placeholder?.key === key)?.placeholder?.default ||
        `<${key}>`
      );
    });
  };

  const [showRenderingCanvas, setShowRenderingCanvas] = useState(false);
  const makeScreenshot = async () => {
    setShowRenderingCanvas(true);
    toast?.(t('toast.generating_screenshot'));
    const toCanvas = await import('html2canvas-add-mix-blend-mode').then(
      (module) => module.default
    );
    const resultsBox = document.getElementById('results');

    if (resultsBox) {
      const canvas = await toCanvas(resultsBox, { backgroundColor: 'transparent' });
      setShowRenderingCanvas(false);
      return await new Promise<Blob>((resolve, reject) =>
        canvas.toBlob(
          (blob) => {
            if (blob) {
              resolve(blob);
            } else {
              reject(new Error('Image not generated'));
            }
          },
          'image/png',
          1
        )
      );
    }
  };

  const screenshot = async () => {
    const shareImage = await makeScreenshot();
    if (!shareImage) return;
    try {
      await navigator.share({
        text: t('share.copy_text'),
        files: [new File([shareImage], 'll-sorted.png')]
      });
    } catch {
      await navigator.clipboard.write([
        new ClipboardItem({ 'image/png': shareImage }, { presentationStyle: 'attachment' })
      ]);
    }
  };

  const download = async () => {
    try {
      const blob = await makeScreenshot();
      if (!blob) return;
      const saveAs = (await import('file-saver')).saveAs;
      saveAs(new File([blob], `${replacePlaceholder(preset.text, placeholderData)}.png`));
    } catch (error) {
      console.log(error);
    }
  };

  return (
    <>
      <Metadata title={title} helmet />
      <Stack alignItems="center" w="full">
        <RadioGroup.Root
          defaultValue={`0`}
          value={`${presetIndex}`}
          onValueChange={({ value }) => setPresetIndex(Number(value))}
        >
          {PRESETS.map((option, index) => {
            const text = replacePlaceholder(option.text, placeholderData);
            return (
              <RadioGroup.Item key={index} value={`${index}`}>
                <RadioGroup.ItemControl />
                <RadioGroup.ItemText>{text}</RadioGroup.ItemText>
                <RadioGroup.ItemHiddenInput />
              </RadioGroup.Item>
            );
          })}
        </RadioGroup.Root>
        {placeholders.map(({ placeholder }) => {
          if (!placeholder) return null;
          return (
            <Input
              key={placeholder.key}
              value={placeholderData?.[placeholder.key]}
              defaultValue={placeholder.default}
              onChange={(e) =>
                setPlaceholderData((d) => ({ ...d, [placeholder.key]: e.target.value }))
              }
            />
          );
        })}
        <Template preset={preset} baseSize={50} placeholderData={placeholderData} />
      </Stack>
      <HStack justifyContent="center">
        <Button variant="subtle" onClick={() => void screenshot()}>
          <FaCopy /> {t('results.copy')}
        </Button>
        <Button onClick={() => void download()}>
          <FaDownload /> {t('results.download')}
        </Button>
      </HStack>
      {showRenderingCanvas && (
        <Box position="absolute" w="0" h="0" overflow="hidden">
          <Stack id="results" w="fit-content" h="fit-content">
            <Template preset={preset} placeholderData={placeholderData} transparent />
          </Stack>
        </Box>
      )}
    </>
  );
}
