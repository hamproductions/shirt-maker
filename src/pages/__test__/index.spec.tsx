import { describe, expect, it } from 'vitest';
import { Page } from '../index/+Page';
import { render } from '~/__test__/utils';

describe('Home page', () => {
  it('renders', async () => {
    const [{ container }] = await render(<Page />);
    expect(container).toMatchSnapshot();
  });
});
