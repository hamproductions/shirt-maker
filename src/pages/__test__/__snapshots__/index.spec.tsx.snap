// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home page > renders 1`] = `
<div>
  <script
    lang="js"
  >
    
            const savedSettings = localStorage.getItem('color-mode')
            if (savedSettings !== null) {
              document.documentElement.classList.add(savedSettings === '"dark"' ? 'dark': 'light');
            } else if (window.matchMedia && window.matchMedia('(prefers-color-scheme: dark)').matches) {
              document.documentElement.classList.add('dark');
              localStorage.setItem("color-mode", '"dark"')
            } 
          
  </script>
  <div
    class="d_flex flex-d_column gap_10px pos_relative w_full min-h_100vh bg-c_bg.default"
  >
    <div
      class="pos_relative max-w_8xl mx_auto px_4 z_1 flex_1 w_full py_4"
    >
      <div
        class="d_flex flex-d_column gap_10px"
      >
        <div
          class="d_flex ai_center jc_flex-end gap_10px flex-d_row w_full"
        >
          <div
            class="d_flex flex-wrap_wrap gap_10px"
          >
            <button
              class="button button--variant_link button--size_md active:fw_bold"
              data-active="true"
            >
              English
            </button>
            |
            <button
              class="button button--variant_link button--size_md active:fw_bold"
            >
              日本語
            </button>
          </div>
          <button
            aria-label="Toggle Color Mode"
            class="button button--variant_subtle button--size_md px_0"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M361.5 1.2c5 2.1 8.6 6.6 9.6 11.9L391 121l107.9 19.8c5.3 1 9.8 4.6 11.9 9.6s1.5 10.7-1.6 15.2L446.9 256l62.3 90.3c3.1 4.5 3.7 10.2 1.6 15.2s-6.6 8.6-11.9 9.6L391 391 371.1 498.9c-1 5.3-4.6 9.8-9.6 11.9s-10.7 1.5-15.2-1.6L256 446.9l-90.3 62.3c-4.5 3.1-10.2 3.7-15.2 1.6s-8.6-6.6-9.6-11.9L121 391 13.1 371.1c-5.3-1-9.8-4.6-11.9-9.6s-1.5-10.7 1.6-15.2L65.1 256 2.8 165.7c-3.1-4.5-3.7-10.2-1.6-15.2s6.6-8.6 11.9-9.6L121 121 140.9 13.1c1-5.3 4.6-9.8 9.6-11.9s10.7-1.5 15.2 1.6L256 65.1 346.3 2.8c4.5-3.1 10.2-3.7 15.2-1.6zM160 256a96 96 0 1 1 192 0 96 96 0 1 1 -192 0zm224 0a128 128 0 1 0 -256 0 128 128 0 1 0 256 0z"
              />
            </svg>
          </button>
        </div>
        <div
          class="d_flex ai_center jc_center"
        >
          <div
            class="d_flex flex-d_column ai_center gap_10px w_full max-w_breakpoint-lg"
          >
            <h1
              class="text text--variant_heading fs_2xl"
            >
              Custom Shirt Maker
            </h1>
            <p
              class="text"
            >
              Want your own custom design? Try it here!
               
              <a
                class="link"
                href="https://x.com/luseo29/status/1799676661627252779"
                target="_blank"
              >
                @luseo29's design.
              </a>
            </p>
            <h2
              class="text text--variant_heading fs_lg"
            >
              Select Design
            </h2>
            <div
              aria-labelledby="radio-group::r0::label"
              aria-orientation="vertical"
              class="radioGroup__root radioGroup__root--size_sm"
              data-orientation="vertical"
              data-part="root"
              data-scope="radio-group"
              dir="ltr"
              id="radio-group::r0:"
              role="radiogroup"
              style="position: relative;"
            >
              <label
                class="radioGroup__item radioGroup__item--size_sm"
                data-orientation="vertical"
                data-part="item"
                data-scope="radio-group"
                data-state="checked"
                dir="ltr"
                for="radio-group::r0::radio:input:0"
                id="radio-group::r0::radio:0"
              >
                <div
                  aria-hidden="true"
                  class="radioGroup__itemControl radioGroup__itemControl--size_sm"
                  data-orientation="vertical"
                  data-part="item-control"
                  data-scope="radio-group"
                  data-state="checked"
                  dir="ltr"
                  id="radio-group::r0::radio:control:0"
                />
                <span
                  class="radioGroup__itemText radioGroup__itemText--size_sm"
                  data-orientation="vertical"
                  data-part="item-text"
                  data-scope="radio-group"
                  data-state="checked"
                  dir="ltr"
                  id="radio-group::r0::radio:label:0"
                >
                  日本語は全て女性声優から学んだ
                </span>
                <input
                  checked=""
                  data-ownedby="radio-group::r0:"
                  id="radio-group::r0::radio:input:0"
                  name=":r0:"
                  style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap; word-wrap: normal;"
                  type="radio"
                  value="0"
                />
              </label>
              <label
                class="radioGroup__item radioGroup__item--size_sm"
                data-orientation="vertical"
                data-part="item"
                data-scope="radio-group"
                data-state="unchecked"
                dir="ltr"
                for="radio-group::r0::radio:input:1"
                id="radio-group::r0::radio:1"
              >
                <div
                  aria-hidden="true"
                  class="radioGroup__itemControl radioGroup__itemControl--size_sm"
                  data-orientation="vertical"
                  data-part="item-control"
                  data-scope="radio-group"
                  data-state="unchecked"
                  dir="ltr"
                  id="radio-group::r0::radio:control:1"
                />
                <span
                  class="radioGroup__itemText radioGroup__itemText--size_sm"
                  data-orientation="vertical"
                  data-part="item-text"
                  data-scope="radio-group"
                  data-state="unchecked"
                  dir="ltr"
                  id="radio-group::r0::radio:label:1"
                >
                  日本語は全て女性声優のため学んだ
                </span>
                <input
                  data-ownedby="radio-group::r0:"
                  id="radio-group::r0::radio:input:1"
                  name=":r0:"
                  style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap; word-wrap: normal;"
                  type="radio"
                  value="1"
                />
              </label>
              <label
                class="radioGroup__item radioGroup__item--size_sm"
                data-orientation="vertical"
                data-part="item"
                data-scope="radio-group"
                data-state="unchecked"
                dir="ltr"
                for="radio-group::r0::radio:input:2"
                id="radio-group::r0::radio:2"
              >
                <div
                  aria-hidden="true"
                  class="radioGroup__itemControl radioGroup__itemControl--size_sm"
                  data-orientation="vertical"
                  data-part="item-control"
                  data-scope="radio-group"
                  data-state="unchecked"
                  dir="ltr"
                  id="radio-group::r0::radio:control:2"
                />
                <span
                  class="radioGroup__itemText radioGroup__itemText--size_sm"
                  data-orientation="vertical"
                  data-part="item-text"
                  data-scope="radio-group"
                  data-state="unchecked"
                  dir="ltr"
                  id="radio-group::r0::radio:label:2"
                >
                  私が日本に来た理由は女性声優を会うために
                </span>
                <input
                  data-ownedby="radio-group::r0:"
                  id="radio-group::r0::radio:input:2"
                  name=":r0:"
                  style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap; word-wrap: normal;"
                  type="radio"
                  value="2"
                />
              </label>
            </div>
            <h2
              class="text text--variant_heading fs_lg"
            >
              Customize Text
            </h2>
            <p
              class="text"
            >
              Switch light/dark mode for black/white text version.
            </p>
            <input
              class="input input--size_md"
              value="女性声優"
            />
            <h2
              class="text text--variant_heading fs_lg"
            >
              Preview
            </h2>
            <canvas
              style="display: block;"
            />
            <div
              class="d_flex ai_center jc_center gap_10px flex-d_row"
            >
              <button
                class="button button--variant_subtle button--size_md"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 448 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M208 0H332.1c12.7 0 24.9 5.1 33.9 14.1l67.9 67.9c9 9 14.1 21.2 14.1 33.9V336c0 26.5-21.5 48-48 48H208c-26.5 0-48-21.5-48-48V48c0-26.5 21.5-48 48-48zM48 128h80v64H64V448H256V416h64v48c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V176c0-26.5 21.5-48 48-48z"
                  />
                </svg>
                 
                Copy to clipboard
              </button>
              <button
                class="button button--variant_solid button--size_md"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 512 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M288 32c0-17.7-14.3-32-32-32s-32 14.3-32 32V274.7l-73.4-73.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l128 128c12.5 12.5 32.8 12.5 45.3 0l128-128c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L288 274.7V32zM64 352c-35.3 0-64 28.7-64 64v32c0 35.3 28.7 64 64 64H448c35.3 0 64-28.7 64-64V416c0-35.3-28.7-64-64-64H346.5l-45.3 45.3c-25 25-65.5 25-90.5 0L165.5 352H64zm368 56a24 24 0 1 1 0 48 24 24 0 1 1 0-48z"
                  />
                </svg>
                 
                Download
              </button>
            </div>
          </div>
        </div>
        <div
          class="pos_absolute w_0 h_0 ov_hidden"
        >
          <div
            class="d_flex flex-d_column gap_10px w_fit-content h_fit-content"
            id="results"
          >
            <canvas
              style="display: block;"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="d_flex flex-d_column jc_center gap_1 w_full p_4 ta_center bg-c_bg.muted"
    >
      <div
        class="d_flex flex-wrap_wrap jc_center gap_10px w_full"
      >
        <p
          class="text"
        >
          Created by
           
          <a
            class="link"
            href="https://ham-san.net/namecard"
            target="_blank"
          >
            ハムP
          </a>
        </p>
         
        | 
        <p
          class="text"
        >
          Thanks @luseo29 for the template
        </p>
      </div>
      <div
        class="d_flex flex-wrap_wrap jc_center gap_10px w_full"
      >
        <p
          class="text"
        >
          Check out source code on 
        </p>
        <a
          class="link"
          href="https://github.com/hamproductions/shirt-maker"
          target="_blank"
        >
          GitHub
        </a>
      </div>
    </div>
    <div
      class="z_0 pos_fixed top_0 left_0 w_100vw h_100vh op_0.05 bg-p_center bg-a_fixed bg-i_var(--bg-image) bg-s_cover mix-bm_darken dark:mix-bm_lighten pointer-events_none"
    />
  </div>
  <div
    aria-label="bottom-end Notifications alt+T"
    aria-live="polite"
    data-align="end"
    data-part="group"
    data-placement="bottom-end"
    data-scope="toast"
    data-side="bottom"
    dir="ltr"
    id="toast-group:bottom-end"
    role="region"
    style="position: fixed; pointer-events: none; display: flex; flex-direction: column; --gap: 16px; --first-height: 0px; z-index: 2147483647; align-items: flex-end; inset-inline-end: calc(env(safe-area-inset-right, 0px) + 1rem);"
    tabindex="-1"
  />
</div>
`;
